cryptoki/tests/basic.rs

```diff
@@ -781,3 +783,270 @@ fn ro_rw_session_test() -> Result<()> {
+    assert_eq!(expected_labels, labels);
+
+    Ok(())
+}
+
+// Find objects matching a template when none match.
+#[test]
+#[serial]
+fn find_objects_none() -> Result<()> {
+    let (pkcs11, slot) = init_pins();
+    let mut session = pkcs11.open_rw_session(slot)?;
+    session.login(UserType::User, Some(USER_PIN))?;
+
+    // Generate a sample object labeled "foo"
+    generate_sample_objects(&session, ["foo"])?;
+
+    // Search for objects labeled "bar"
+    let objects = session.find_objects(&[Attribute::Label(b"bar".to_vec())])?;
+    assert_eq!(&objects, &[]);
+    Ok(())
+}
+
+#[test]
+#[serial]
+// Find objects matching a template when a few (<10) match.
+fn find_objects_few() -> Result<()> {
+    let (pkcs11, slot) = init_pins();
+    let mut session = pkcs11.open_rw_session(slot)?;
+    session.login(UserType::User, Some(USER_PIN))?;
+
+    // generate some sample AES keys to match
+    let expected_labels = vec!["bar", "baz", "foo"];
+    generate_sample_objects(&session, &expected_labels)?;
+
+    // generate a key that shouldn't match (DES3 vs AES)
+    session.generate_key(
+        &Mechanism::Des3KeyGen,
+        &[Attribute::Label(b"quux".to_vec())],
+    )?;
+
+    // search for all AES keys
+    let objects = session.find_objects(&[Attribute::KeyType(KeyType::AES)])?;
+    let labels = object_labels_sorted(&session, objects)?;
+    assert_eq!(expected_labels, labels);
+    Ok(())
+}
+
+// Find objects matching a template when many (>10) match.
+#[test]
+#[serial]
+fn find_objects_many() -> Result<()> {
+    let (pkcs11, slot) = init_pins();
+    let mut session = pkcs11.open_rw_session(slot)?;
+    session.login(UserType::User, Some(USER_PIN))?;
+
+    // generate some sample AES keys to match
+    let expected_labels = (1..=20)
+        .map(|x| format!("key {:02}", x))
+        .collect::<Vec<String>>();
+    generate_sample_objects(&session, &expected_labels)?;
+
+    // generate a key that shouldn't match (DES3 vs AES)
+    session.generate_key(
+        &Mechanism::Des3KeyGen,
+        &[Attribute::Label(b"quux".to_vec())],
+    )?;
+
+    // search for all AES keys
+    let objects = session.find_objects(&[Attribute::KeyType(KeyType::AES)])?;
+    let labels = object_labels_sorted(&session, objects)?;
+    assert_eq!(expected_labels, labels);
+    Ok(())
+}
+
+// Find all objects incrementally (i.e. empty template)
+#[test]
+#[serial]
+fn find_objects_incrementally_all() -> Result<()> {
+    let (pkcs11, slot) = init_pins();
+    let mut session = pkcs11.open_rw_session(slot)?;
+    session.login(UserType::User, Some(USER_PIN))?;
+
+    // Generate some sample objects
+    let expected_labels = vec!["bar", "baz", "foo"];
+    generate_sample_objects(&session, &expected_labels)?;
+
+    // Find all objects
+    let mut search = session.find_objects_init(&[])?;
+    let objects = search.find_next(10)?;
+    assert!(search.find_next(10)?.is_empty());
+    drop(search);
```

Comment:
Hmmmmm... Would it be possible to add a method to `FindObjects`, something like `finalize`, that just takes ownership of the value and lets it drop? It seems suboptimal to have to drop the object manually to achieve this, and an empty method isn't too bad.